Cleanup/drop machine access

blackboxsw opened the pull request:
First part is to read and process service access information directly from machine-token instead of specific machine-access file.

Branch items:
 *  UAConfig.entitlements understands how to read resoureTokens list from machine-token
 * contract.request_updated_contract now pulls delta information from cfg.entitlements object
 * UAContractClient.request_machine_token_update clears cfg._entitlements cache

 The second branch drop UAContractClient.request_resource_machine_access and any callsites.

OddBloke commented on uaclient/entitlements/tests/conftest.py:
```diff
@@ -3,51 +3,68 @@
     *,
     affordances: "Dict[str, Any]" = None,
     directives: "Dict[str, Any]" = None,
+    entitled: "Optional[bool]" = True,
```
I don't think this is `Optional`; it should only ever be `True` or `False`, so it's a straight-up `bool`. I just confirmed this for myself with a toy example:

```py
from typing import Optional


def optional_foo(x: Optional[bool] = True) -> None:
    pass


optional_foo(True)
optional_foo(False)
optional_foo()
optional_foo(None)


def foo(x: bool = True) -> None:
    pass


foo(True)
foo(False)
foo()
foo(None)
```

gives:

```
$ mypy foo.py
foo.py:21: error: Argument 1 to "foo" has incompatible type "None"; expected "bool"
```

blackboxsw replied:
oopsie. got excited about the Optional. Dropped

OddBloke commented on uaclient/entitlements/tests/conftest.py:
```diff
@@ -3,51 +3,68 @@
-        },
+        "obligations": obligations,
+        "type": type,
+        "entitled": True,
```
Should this be using the value of `entitled`?

blackboxsw replied:
yes, nice one

OddBloke commented on uaclient/entitlements/tests/test_cc.py:
```diff
@@ -168,7 +160,7 @@ def exists(path):
                 "/etc/apt/sources.list.d/ubuntu-cc-eal-xenial.list",
                 "http://CC",
-                "TOKEN",
+                "%s-token" % entitlement.name,
```
A nit, but `.format` here would be more consistent with the rest of the codebase.

OddBloke commented on uaclient/entitlements/tests/test_cis.py:
```diff
@@ -63,7 +63,7 @@ def fake_platform(key=None):
                 "/etc/apt/sources.list.d/ubuntu-cis-audit-xenial.list",
                 "http://CIS-AUDIT",
-                "TOKEN",
+                "%s-token" % entitlement.name,
```
A nit, but `.format` here would be more consistent with the rest of the codebase.

OddBloke commented on uaclient/entitlements/tests/test_esm.py:
```diff
@@ -73,7 +73,7 @@ def fake_exists(path):
                 ),
                 "http://{}".format(entitlement.name.upper()),
-                "TOKEN",
+                "%s-token" % entitlement.name,
```
A nit, but `.format` here would be more consistent with the rest of the codebase.

OddBloke commented on uaclient/entitlements/tests/test_esm.py:
```diff
@@ -157,7 +157,7 @@ def fake_subp(cmd, capture=None, retry_sleeps=None):
                 ),
                 "http://{}".format(entitlement.name.upper()),
-                "TOKEN",
+                "%s-token" % entitlement.name,
```
A nit, but `.format` here would be more consistent with the rest of the codebase.

OddBloke commented on uaclient/entitlements/tests/test_fips.py:
```diff
@@ -88,7 +87,7 @@ def test_enable_configures_apt_sources_and_auth_files(self, entitlement):
                 ),
                 repo_url,
-                "TOKEN",
+                "%s-token" % entitlement.name,
```
A nit, but `.format` here would be more consistent with the rest of the codebase.

OddBloke commented on uaclient/entitlements/tests/conftest.py:
```diff
@@ -3,51 +3,68 @@
+    suites: "List[str]" = None
+) -> "Dict[str, Any]":
     return {
+        "resourceTokens": [{"Type": type, "Token": "%s-token" % type}],
```
A nit, but `.format` here would be more consistent with the rest of the codebase.

OddBloke commented on uaclient/contract.py:
```diff
@@ -280,18 +282,11 @@ def request_updated_contract(
-    for name, entitlement in sorted(cfg.entitlements.items()):
-        if entitlement["entitlement"].get("entitled"):
-            # Obtain each entitlement's accessContext for this machine
-            new_access = contract_client.request_resource_machine_access(
```
Have we confirmed with the contract server team that they are prepared for any consequences of us not hitting this endpoint any longer?

blackboxsw replied:
We had in IRC with mhilton and cmars before starting this work. And their consensus was that the can (and should be) determinig KPI attach metrics from their backend service endpoints (like livepatch of various PPA paths during apt-update calls).  We may have to restate that we have completed this effort again to confirm that they aren't relying on information from ua client attaching via those routes.

blackboxsw replied:
pinged in #ua IRC again about this cmars will chat with the team tomorrow.

blackboxsw replied:
confirmed in IRC today from alnvdl and cmars that this is not an issue.

OddBloke commented on uaclient/config.py:
```diff
@@ -115,19 +115,19 @@ def entitlements(self):
         contractInfo = machine_token["machineTokenInfo"]["contractInfo"]
-
+        tokens_by_name = dict(
+            (e["Type"], e["Token"])
```
So this part of the response has the `Type` key capitalised, but the one we use in the next logical line has it lower-case? Is consistency something we could ask of ua-contracts, or has the ship sailed on changing the shape of the API response?

blackboxsw replied:
I think it's worth a bug against ua-contracts on that, since we are the only consumer of that data. We can make this branch work with the lower() value in either case.

blackboxsw replied:
Filed https://github.com/CanonicalLtd/ua-contracts/issues/645

OddBloke commented on uaclient/entitlements/tests/conftest.py:
```diff
@@ -3,51 +3,68 @@
+    *,
+    affordances: "Dict[str, Any]" = None,
+    directives: "Dict[str, Any]" = None,
+    entitled: "Optional[bool]" = True,
```
As I mention in more detail below (I missed this definition on my first read through), this isn't `Optional`.

blackboxsw replied:
dropped to bool

blackboxsw commented on uaclient/contract.py:
```diff
@@ -174,6 +174,8 @@ def request_machine_token_update(
         if headers.get("expires"):
             response["expires"] = headers["expires"]
+        # Clear UAConfig._entitlements cache because we are refreshing
+        self.cfg.delete_cache_key("machine-token")
```
ahh right, and turns out it already does, this was a shortcoming in our FakeConfig which doesn't delete that cache key during a FakeConfig.write_cache call.

blackboxsw replied:
Just refactored FakeConfig subclass to require a data_dir param (presumably tmpdir.strpath) and updated all unit tests. Our FakeConfig was growing too much parallel logic with existing UAConfig and that was causing the unit test failure which prompted me to incorrectly add the "delete_cache_key" call.

blackboxsw commented on uaclient/entitlements/tests/conftest.py:
```diff
@@ -3,51 +3,68 @@
 
 def machine_access(
-    entitlement_type: str,
+    type: str,
```
done

OddBloke commented on uaclient/entitlements/tests/test_livepatch.py:
```diff
@@ -51,9 +51,15 @@ def test_contract_status_entitled(self, entitlement):
         """The contract_status returns NONE when entitled is False."""
-        entitlement.cfg.write_cache(
-            "machine-access-livepatch", {"entitlement": {"entitled": False}}
+        token = entitlement.cfg.read_cache("machine-token")
```
I haven't fully thought this through, so please take it with a grain of salt, but might these test changes be simpler if you used `entitlement_factory` directly, rather than using the `entitlement` fixture and then modifying it after the fact? (Looking at the `entitlement` fixture, it might make sense to introduce a `livepatch_entitlement_factory` fixture, which sets affordances/directives in the same way as `entitlement`.)

Something like (completely untested):

```py
@pytest.fixture
def livepatch_entitlement_factory(entitlement_factory):

    def factory_func(entitled: bool = True, suites: List[str] = None):
        affordances = {
            "architectures": ["x86_64"],
            "minKernelVersion": "4.4",
            "kernelFlavors": ["generic", "lowlatency"],
            "tier": "stable",
        }
        directives = {"caCerts": "", "remoteServer": "https://alt.livepatch.com"}
        return entitlement_factory(
            LivepatchEntitlement, affordances=affordances, directives=directives,
            entitled=entitled, suites=suites,
        )

    return factory_func

@pytest.fixture
def entitlement(livepatch_entitlement_factory):
    return livepatch_entitlement_factory()
```

and then this test would be something like:

```py
    def test_contract_status_unentitled(self, livepatch_entitlement_factory):
        entitlement = livepatch_entitlement_factory(entitled=False)
        assert ContractStatus.UNENTITLED == entitlement.contract_status()
```

(This would require adding an `entitled` argument to `entitlement_factory` too, and passing it through to `machine_token` in `conftest.py`.)

blackboxsw replied:
> I haven't fully thought this through, so please take it with a grain of salt, but might these test changes be simpler if you used `entitlement_factory` directly, rather than using the `entitlement` fixture and then modifying it after the fact? (Looking at the `entitlement` fixture, it might make sense to introduce a `livepatch_entitlement_factory` fixture, which sets affordances/directives in the same way as `entitlement`.)
> 
> Something like (completely untested):
> 
> ```python
> @pytest.fixture
> def livepatch_entitlement_factory(entitlement_factory):
> 
>     def factory_func(entitled: bool = True, suites: List[str] = None):
>         affordances = {
>             "architectures": ["x86_64"],
>             "minKernelVersion": "4.4",
>             "kernelFlavors": ["generic", "lowlatency"],
>             "tier": "stable",
>         }
>         directives = {"caCerts": "", "remoteServer": "https://alt.livepatch.com"}
>         return entitlement_factory(
>             LivepatchEntitlement, affordances=affordances, directives=directives,
>             entitled=entitled, suites=suites,
>         )
> 
>     return factory_func
> 
> @pytest.fixture
> def entitlement(livepatch_entitlement_factory):
>     return livepatch_entitlement_factory()
> ```
> 
> and then this test would be something like:
> 
> ```python
>     def test_contract_status_unentitled(self, livepatch_entitlement_factory):
>         entitlement = livepatch_entitlement_factory(entitled=False)
>         assert ContractStatus.UNENTITLED == entitlement.contract_status()
> ```
> 
> (This would require adding an `entitled` argument to `entitlement_factory` too, and passing it through to `machine_token` in `conftest.py`.)

I like this suggestion a lot, that whole logic I added in testing was messy and felt dirty. Thanks for the thoughts on this will implement tomorrow.

OddBloke commented on uaclient/config.py:
```diff
@@ -115,19 +115,21 @@ def entitlements(self):
+        tokens_by_name = dict(
+            # Handle Upper-case and lower-case keys until the following issue
+            # is fixed: ua-contracts #645
+            (e.get("Type", e.get("type")), e.get("Token", e.get("token")))
```
I don't want to land this branch with this logic in place. We should work out what the contract server should be doing, get them to make the changes, and write against that.

blackboxsw commented:
@OddBloke addressed all comments. Also validated that contracts.staging has been fixes to use lowercase `type` and `token` keys in `resourceTokens` in the machine-token.json. Still can't land/release this until production has this fix so I'm leaving the external blocker label on it until I can confirm a production fix.

OddBloke commented on uaclient/tests/test_contract.py:
```diff
@@ -384,26 +378,20 @@ def test_attached_config_refresh_machine_token_and_services(
         }
+        new_token = copy.deepcopy(machine_token)
+        new_token["machineTokenInfo"]["contractInfo"]["resourceEntitlements"][
+            1
```
Haha, thanks black!

blackboxsw replied:
yes glad I don't have to worry about defending automated formatting. as this would have been a couple of cycles. I guess I could better construct locals to avoid that nastiness. But it's more fun to don't care

OddBloke replied:
Oh yeah, this is fine, it's totally worth the occasional dumb thing like this to not have to think about it at all the rest of the time.

OddBloke commented on uaclient/entitlements/tests/test_fips.py:
```diff
@@ -143,14 +142,15 @@ def test_enable_returns_false_on_missing_suites_directive(
+        token["machineTokenInfo"]["contractInfo"]["resourceEntitlements"] = [
+            entitlement_cfg
+        ]
+        entitlement.cfg.write_cache("machine-token", token)
```
Similar to the other test we discussed, would this be simpler by using an entitlement factory directly? I think so:

```diff
diff --git a/uaclient/entitlements/tests/test_fips.py b/uaclient/entitlements/tests/test_fips.py
index a44672e..6f699d5 100644
--- a/uaclient/entitlements/tests/test_fips.py
+++ b/uaclient/entitlements/tests/test_fips.py
@@ -4,6 +4,7 @@ import contextlib
 import copy
 import itertools
 import mock
+from functools import partial
 
 import pytest
 
@@ -24,9 +25,14 @@ M_GETPLATFORM = M_REPOPATH + "util.get_platform_info"
 
 
 @pytest.fixture(params=[FIPSEntitlement, FIPSUpdatesEntitlement])
-def entitlement(request, entitlement_factory):
+def fips_entitlement_factory(request, entitlement_factory):
     """Parameterized fixture so we apply all tests to both FIPS and Updates"""
-    return entitlement_factory(request.param)
+    return partial(entitlement_factory, request.param)
+
+
+@pytest.fixture
+def entitlement(fips_entitlement_factory):
+    return fips_entitlement_factory()
 
 
 class TestFIPSEntitlementCanEnable:
@@ -138,19 +144,10 @@ class TestFIPSEntitlementEnable:
         "uaclient.util.get_platform_info", return_value={"series": "xenial"}
     )
     def test_enable_returns_false_on_missing_suites_directive(
-        self, m_platform_info, m_add_apt, entitlement
+        self, m_platform_info, m_add_apt, fips_entitlement_factory
     ):
         """When directives do not contain suites returns false."""
-        # Unset suites directive
-        token = entitlement.cfg.read_cache("machine-token")
-        entitlement_cfg = entitlement.cfg.entitlements[entitlement.name].get(
-            "entitlement"
-        )
-        entitlement_cfg["directives"]["suites"] = []
-        token["machineTokenInfo"]["contractInfo"]["resourceEntitlements"] = [
-            entitlement_cfg
-        ]
-        entitlement.cfg.write_cache("machine-token", token)
+        entitlement = fips_entitlement_factory(suites=[])
 
         with mock.patch.object(entitlement, "can_enable", return_value=True):
             with pytest.raises(exceptions.UserFacingError) as excinfo:
```

OddBloke replied:
(Oh, FYI, this patch won't apply cleanly without first applying the one above.)

OddBloke commented on uaclient/entitlements/tests/test_livepatch.py:
```diff
@@ -29,31 +36,51 @@
-    return entitlement_factory(
-        LivepatchEntitlement, affordances=affordances, directives=directives
-    )
+def livepatch_entitlement_factory(entitlement_factory):
```
When writing the fips equivalent to this, I realised we could make this simpler with `functools.partial` too:

```diff
diff --git a/uaclient/entitlements/tests/test_livepatch.py b/uaclient/entitlements/tests/test_livepatch.py
index 0e36780..26872c7 100644
--- a/uaclient/entitlements/tests/test_livepatch.py
+++ b/uaclient/entitlements/tests/test_livepatch.py
@@ -3,6 +3,7 @@
 import copy
 import logging
 import mock
+from functools import partial
 from types import MappingProxyType
 
 try:
@@ -46,26 +47,13 @@ DEFAULT_AFFORDANCES = {
 
 @pytest.fixture
 def livepatch_entitlement_factory(entitlement_factory):
-    def factory_func(
-        entitled: bool = True,
-        affordances: "Dict[str, Any]" = None,
-        suites: "List[str]" = None,
-    ):
-        if not affordances:
-            affordances = DEFAULT_AFFORDANCES
-        directives = {
-            "caCerts": "",
-            "remoteServer": "https://alt.livepatch.com",
-        }
-        return entitlement_factory(
-            LivepatchEntitlement,
-            affordances=affordances,
-            directives=directives,
-            entitled=entitled,
-            suites=suites,
-        )
-
-    return factory_func
+    directives = {"caCerts": "", "remoteServer": "https://alt.livepatch.com"}
+    return partial(
+        entitlement_factory,
+        LivepatchEntitlement,
+        affordances=DEFAULT_AFFORDANCES,
+        directives=directives,
+    )
 
 
 @pytest.fixture
```

blackboxsw commented:
Thanks dan for the patches, I've applied all 3.
